### PR TITLE
Fix cross join delay condition in Greedy algorithm

### PR DESCRIPTION
**Pull request by asubramanya**
Earlier, we were relying on the number of edges to identify if we have
potential opportunities to create joins avoiding cross joins.  However, we
should actually iterate over all the edges to find overlapping edges. If any
edge exists which is overlapping with the current join, then we should avoid
creating a cross join with other component which donot have an overlapping edge.

Co-authored-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>

### CODE REVIEW

File: libgpopt/src/xforms/CJoinOrderGreedy.cpp [hunk 1]
```diff
@@ -215,11 +215,37 @@ CJoinOrderGreedy::PexprExpand()
 
 			// combine component with current result and derive stats
 			CJoinOrder::SComponent *pcompTemp = PcompCombine(m_pcompResult, pcompCurrent);
-			if(CUtils::FCrossJoin(pcompTemp->m_pexpr) && 1 < m_ulNumUsedEdges)
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrderGreedy.cpp]: Please rename m_ulNumUsedEdges. It is confusing.

File: libgpopt/include/gpopt/xforms/CJoinOrderGreedy.h [hunk 2]
```diff
@@ -35,7 +35,7 @@ namespace gpopt
 			// result component
 			SComponent *m_pcompResult;
 
-			ULONG m_ulNumUsedEdges;
+			ULONG m_ulNumEdgesRemaining;
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/xforms/CJoinOrderGreedy.h]: comment please

File: libgpopt/src/xforms/CJoinOrderGreedy.cpp [hunk 3]
```diff
@@ -185,65 +184,67 @@ CJoinOrderGreedy::PexprExpand()
...
-		m_pcompResult->Release();
-		m_pcompResult = pcompBestResult;
+		else
+		{
+			// only cross joins are available. pick the unused component which will
+			// result in minimal cardinality
+			best_comp_idx = PickBestJoin(unused_components_set);
+			unused_components_set->ExchangeClear(best_comp_idx);
+		}
+		candidate_comp_set->Release();
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply libgpopt/src/xforms/CJoinOrderGreedy.cpp]: Added assert.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 4]
```diff
@@ -301,8 +306,10 @@ CJoinOrder::ComputeEdgeCover()
 			CExpression *pexprComp = m_rgpcomp[ulComp]->m_pexpr;
 			CColRefSet *pcrsOutput = CDrvdPropRelational::GetRelationalProperties(pexprComp->PdpDerive())->PcrsOutput();
 
+
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply libgpopt/src/xforms/CJoinOrder.cpp]: Updated comment.

File: libgpopt/src/xforms/CJoinOrderGreedy.cpp [hunk 5]
```diff
@@ -185,65 +184,68 @@ CJoinOrderGreedy::PexprExpand()
...
 
 	if(NULL != m_pcompResult)
 	{
-		ulCoveredComps = 2;
+		// found atleast one non cross join
 		MarkUsedEdges();
 	}
 	else
 	{
+		// if every join combination is a cross join
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrderGreedy.cpp]:
```suggestion
		// every join combination is a cross join
```

File: libgpopt/src/xforms/CJoinOrderGreedy.cpp [hunk 6]
```diff
@@ -185,65 +184,68 @@ CJoinOrderGreedy::PexprExpand()
...
 		m_pcompResult = GPOS_NEW(m_mp) SComponent(m_mp, NULL /*pexpr*/);
 	}
-
-	while (ulCoveredComps < m_ulComps)
+	
+	// create a bitset for all the unused components
+	CBitSet *unused_components_set = GPOS_NEW(m_mp) CBitSet(m_mp);
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrderGreedy.cpp]: Any reason why we can't use `m_ulNumEdgesRemaining`? If there is, add a comment here as to why.

File: libgpopt/src/xforms/CJoinOrderGreedy.cpp [hunk 7]
```diff
@@ -185,65 +184,68 @@ CJoinOrderGreedy::PexprExpand()
...
-				continue;
-			}
+			unused_components_set->ExchangeSet(ul);
+		}
+	}
+	
+	while (unused_components_set->Size() > 0)
+	{
+		// get a set of components which can be joined with m_pcompResult
+		CBitSet *candidate_comp_set = GetJoinCandidateComponents();
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrderGreedy.cpp]: Rename suggestion: GetJoinCandidateComponents -> GetAdjacentComponentsToJoinCandidate. Or GetJoinCandidateAdjacentComponents.

File: libgpopt/src/xforms/CJoinOrderGreedy.cpp (hunk 6 above)
Comment (iyerr3) [reply libgpopt/src/xforms/CJoinOrderGreedy.cpp]: `m_ulNumEdgesRemaining` was replaced with a set of unused components. Hence we don't need that member variable anymore and has been removed in the latest commit.

**Pull request merged**